### PR TITLE
Add DilatedConvolutionConverter, BatchToSpaceFusion and SpaceToBatchF…

### PR DESCRIPTION
**Pull request by mateusztabaka**
…usion transformations

### Tickets:
 - 49797

### CODE REVIEW

**Comment by GlebKazantaev** [comment]
@mateusztabaka could you please add links to MO transformations that perform the same replacement? That would be much easier to review new nGraph transformations.

File: inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    auto starts_pattern = pattern::wrap_type<opset6::Constant>();
+    auto ends_pattern = pattern::wrap_type<opset6::Constant>();
+    auto slice_pattern = pattern::wrap_type<opset6::StridedSlice>({depth_to_space_pattern, starts_pattern, ends_pattern,
+            pattern::wrap_type<opset6::Constant>()});
+    auto reshape_after_pattern = pattern::wrap_type<opset6::Reshape>({slice_pattern, pattern::wrap_type<opset6::Constant>()});
+    auto trans_after_pattern = pattern::wrap_type<opset6::Transpose>({slice_pattern, pattern::wrap_type<opset6::Constant>()});
+    auto reshape_or_transpose_after_pattern = std::make_shared<pattern::op::Or>(OutputVector{reshape_after_pattern, trans_after_pattern});
+
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto& pattern_map = m.get_pattern_value_map();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp:36]:
```suggestion
        cons auto & pattern_map = m.get_pattern_value_map();
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        auto get_reshape_or_transpose = [&pattern_map] (const std::shared_ptr<Node>& reshape_pattern,
+                                                        const std::shared_ptr<Node>& trans_pattern) -> std::shared_ptr<Node> {
+            if (pattern_map.count(reshape_pattern) > 0)
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp:40]:
```suggestion
            if (pattern_map.count(reshape_pattern))
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                return pattern_map.at(reshape_pattern).get_node_shared_ptr();
+            if (pattern_map.count(trans_pattern) > 0)
+                return pattern_map.at(trans_pattern).get_node_shared_ptr();
+            return nullptr;
+        };
+        auto check_input_output_shape = [] (const std::shared_ptr<Node>& node) -> bool {
+            const auto& input_shape = node->get_input_shape(0);
+            const auto& output_shape = node->get_output_shape(0);
+            if (input_shape.size() != 4 || output_shape.size() != 4)
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp:49]: please use pattern predicates to check rank sizes.

File: inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                return false;
+            // Transpose permutation has to be [1, 0, 2, 3]
+            if (!(input_shape[0] == output_shape[1] &&
+                  input_shape[1] == output_shape[0] &&
+                  input_shape[2] == output_shape[2] &&
+                  input_shape[3] == output_shape[3]))
+                return false;
+            return true;
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp:57]:
```suggestion
            return (input_shape[0] == output_shape[1] &&
                  input_shape[1] == output_shape[0] &&
                  input_shape[2] == output_shape[2] &&
                  input_shape[3] == output_shape[3]);
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                std::vector<int64_t>{1, 1, block_size, block_size});
+        auto starts = std::dynamic_pointer_cast<opset6::Constant>(pattern_map.at(starts_pattern).get_node_shared_ptr());
+        if (!starts)
+            return false;
+        auto ends = std::dynamic_pointer_cast<opset6::Constant>(pattern_map.at(ends_pattern).get_node_shared_ptr());
+        if (!ends)
+            return false;
+        auto starts_value = starts->cast_vector<int64_t>();
+        auto ends_value = ends->cast_vector<int64_t>();
+        for (size_t i = 0; i < ends_value.size(); i++) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/batch_to_space_fusion.cpp:90]: Add comments that explain logic below.

File: inference-engine/src/transformations/src/transformations/common_optimizations/dilated_convolution_converter.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    auto pads_end_pattern = pattern::wrap_type<opset6::Constant>();
+    auto space_to_batch_pattern = pattern::wrap_type<opset6::SpaceToBatch>({data_pattern, block_shape_pattern, pads_begin_pattern, pads_end_pattern});
+    auto conv_pattern = pattern::wrap_type<opset6::Convolution>({space_to_batch_pattern, pattern::any_input()});
+    auto crops_begin_pattern = pattern::wrap_type<opset6::Constant>();
+    auto crops_end_pattern = pattern::wrap_type<opset6::Constant>();
+    auto batch_to_space_pattern = pattern::wrap_type<opset6::BatchToSpace>({conv_pattern, pattern::any_input(),
+            crops_begin_pattern, crops_end_pattern});
+
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto pattern_map = m.get_pattern_value_map();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/dilated_convolution_converter.cpp:33]:
```suggestion
        const auto & pattern_map = m.get_pattern_value_map();
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/space_to_batch_fusion.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    auto pads_end_pattern = pattern::wrap_type<opset6::Constant>();
+    auto pad_value = pattern::wrap_type<opset6::Constant>();
+    auto pad_pattern = pattern::wrap_type<opset6::Pad>({reshape_or_transpose_before_pattern, pads_begin_pattern, pads_end_pattern, pad_value});
+    auto space_to_depth_pattern = pattern::wrap_type<opset6::SpaceToDepth>({pad_pattern}, pattern::has_static_shape());
+    auto reshape_after_pattern = pattern::wrap_type<opset6::Reshape>({space_to_depth_pattern, pattern::wrap_type<opset6::Constant>()});
+    auto trans_after_pattern = pattern::wrap_type<opset6::Transpose>({space_to_depth_pattern, pattern::wrap_type<opset6::Constant>()});
+    auto reshape_or_transpose_after_pattern = std::make_shared<pattern::op::Or>(OutputVector{reshape_after_pattern, trans_after_pattern});
+
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto pattern_map = m.get_pattern_value_map();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/space_to_batch_fusion.cpp:36]:
```suggestion
        const auto & pattern_map = m.get_pattern_value_map();
```

**Comment by mateusztabaka** [comment]
> @mateusztabaka could you please add links to MO transformations that perform the same replacement? That would be much easier to review new nGraph transformations.

https://github.com/openvinotoolkit/openvino/blob/11f591c16852637506b1b40d083b450e56d0c8ac/model-optimizer/extensions/middle/DilatedConvolution.py - this is DilatedConvolutionConverter.
Regarding BatchToSpace and SpaceToBatch fusions - since ONNX doesn't support those ops - they are implemented by a pair of Transpose ([1, 0, 2, 3]) with Slice/Pad and DepthToSpace/SpaceToDepth

**Review by postrational — APPROVED** [review]

File: ngraph/core/src/op/convert.cpp [hunk 8]
```diff
@@ -115,6 +116,7 @@ namespace convert
             NGRAPH_TYPE_CASE(evaluate_convert, u64, arg, out);
             NGRAPH_TYPE_CASE(evaluate_convert, f16, arg, out);
             NGRAPH_TYPE_CASE(evaluate_convert, f32, arg, out);
+            NGRAPH_TYPE_CASE(evaluate_convert, boolean, arg, out);
```
Comment (ilyachur) [inline ngraph/core/src/op/convert.cpp:119]: Why do we need to have convert for boolean?

Comment (mateusztabaka) [reply ngraph/core/src/op/convert.cpp:119]: In the model I'm working on, there are subgraphs with Convert ops on booleans that can be easily constant folded. It's a small change so I thought I can include it here.

Comment (ilyachur) [reply ngraph/core/src/op/convert.cpp:119]: Can we add new tests to check that it works as expected?

Comment (mateusztabaka) [reply ngraph/core/src/op/convert.cpp:119]: Tests added

**Review by ilyachur — APPROVED** [review]

File: ngraph/core/include/ngraph/pattern/op/pattern.hpp [hunk 9]
```diff
@@ -64,6 +64,9 @@ namespace ngraph
         NGRAPH_API
         std::function<bool(Output<Node>)> has_static_rank();
 
+        NGRAPH_API
+        std::function<bool(Output<Node>)> rank_equals(Dimension expected_rank);
```
Comment (GlebKazantaev) [inline ngraph/core/include/ngraph/pattern/op/pattern.hpp:68]:
```suggestion
        std::function<bool(Output<Node>)> rank_equals(const Dimension & expected_rank);
```

**Pull request merged**